Restructure build

matrei opened the pull request:
Make the build composition based.

matrei committed:
build(root): move `buildscript` to `buildSrc`

matrei committed:
build(root): align on single quotes for strings

matrei committed:
build(root): remove redundant `project` references

Since `project` is the default delegate, explicit references
are unnecessary and can be removed for cleaner code.

matrei committed:
build(root): simplify getting conditional properties

Since `findProject` returns `null` for missing properties, we
can leverage that to simplify getting and setting properties
that can be provided from multiple sources.

matrei committed:
build(root): simplify `GString`s

Remove unnecessary curly brackets from `GString`s for improved
readability.

matrei committed:
build(root): remove `transitionCheckOptions`

This commit removes the explicit setting of `transitionCheckOption`
for `nexusPublishing`.

The default values are:
- `maxRetries`: 60
- `delayBetween`: 10

matrei committed:
build(root): remove unnecessary `javadoc` task lib exclusion

Remove the explicit exclusion of `javaee-web-api-6.0.jar`
for `groovydoc` and `javadoc` tasks, as it is no longer
on the classpath.

matrei committed:
build(root): remove unnecessary check for insecure repositories

This is handled automatically by Gradle since version 7 unless
`allowInsecureProtocol` is explicitly set to `true`.

matrei committed:
build(root): remove `buildscriptDependencies` task

We are using `buildSrc` now. To list the build dependencies, do:
- `cd buildSrc`
- `gradle dependencies`

matrei committed:
build(root): remove unnecessary `evaluationDependsOn`

This was added a long time ago, but it's unclear to me why.
It can be reverted if it's found to be required later.

matrei committed:
build(root): import `InitializeNexusStagingRepository`

matrei committed:
build(root): align `ext` property setting

matrei committed:
build(root): align on no trailing slash on urls

matrei committed:
build(root): use `configureEach`

matrei committed:
build(root): use `manifest.attributes()` for `Jar` tasks

Replace `manifest.mainAttributes()` with the public Gradle API method
`manifest.attributes()`, avoiding reliance on the internal
`DefaultManifest` class.

matrei committed:
chore: improve code comments in root `build.gradle`

matrei committed:
build(root): remove obsolete Groovy joint workflow code

Eliminate outdated `groovy.jar` resolution, likely from a previous
version of the Groovy joint compilation workflow.

matrei committed:
build(root): use `layout` for file references

Refactor file resolution to use `layout` API for better
compatibility with Gradle's lazy evaluation and configuration
cache.

matrei committed:
build(root): remove unnecessary `async` project handling

The special handling of `async` projects appears to be unnecessary
and can most probably be removed.

matrei committed:
build(root): use string instead of variable

Clarify the code by directly using the `:grails-bom` project name
instead of the variable, as the project name is already established.

matrei committed:
build(root): set Maven credentials directly

Set the Maven repository publishing credentials directly,
avoiding unnecessary use of variables.

matrei committed:
build(root): align usage of `MavenArtefactRepository.setUrl(URI)`

Ensure consistency by aligning with the recommended usage of
`MavenArtifactRepository.setUrl(URI)` for setting repository URLs.

matrei committed:
build(root): cleanup

matrei committed:
build(root): remove unused property `springLoadedCommonOptions`

matrei committed:
build(root): simplify Nexus publishing credentials

- Remove unused properties in `ext`
- Set credentials directly rather than using variables

matrei committed:
build(root): remove transitive Groovy upgrade

The upgrade of transitive Groovy dependencies is no longer necessary.

matrei committed:
build(root): configure Nexus publish extension with type-safe API

Using the type-safe API for configuring the Nexus publish extension provides
IDE type hints and auto-completion.

matrei committed:
build(root): use `projectVersion` to set root project `version`

Set the root project version using `projectVersion` for consistency.

matrei committed:
build(root): refactor `subprojects` iteration

- Remove `subprojects` closure parameter and reference the delegate directly
- Use lazy, type-safe API when configuring extensions
- Replace `ext` properties with variables where possible
- Eliminate unnecessary `uri()`` method in repository URL setters
- Use `=` for setters and parentheses for methods for clarity
- Remove unnecessary empty string default when `null` suffices

matrei committed:
build(root): cast property to `String`

matrei committed:
build(root): remove faulty comment

Remove inaccurate or misleading comment from the build script.

matrei committed:
build(root): remove unnecessary `JavadocFixTool`

The vulnerability this tool was solving has been patched:
https://www.kb.cert.org/vuls/id/225657

matrei committed:
build(root): remove unnecessary empty string default

Remove unnecessary empty string default when `null` suffices

matrei committed:
build(root): consolidate `Javadoc` task configuration

matrei committed:
chore: comments in root `build.gradle`

matrei committed:
build(root): replace `ant.copy` with `sync`

Refactor `ant.copy` usage to `sync`.

matrei committed:
build(root): make helper method static

matrei committed:
build(root): remove legacy IDE Gradle plugins

Remove the obsolete idea and eclipse Gradle plugins,
as modern IDEs handle Gradle projects natively.

matrei committed:
build(root): fix faulty javadoc options

matrei committed:
build(root): cleanup - use parentheses for methods

matrei committed:
build: move `grails-bom` dependency to projects

Each project should explicitly declare its own dependencies
for better clarity and maintainability.

matrei committed:
build: move dependency declaration to projects

Each project should explicitly declare its own dependencies
for better clarity and maintainability.

matrei committed:
build: move dependency declaration to projects

Each project should explicitly declare its own dependencies
for better clarity and maintainability.

matrei committed:
build: move dependency declaration to projects

Each project should explicitly declare its own dependencies
for better clarity and maintainability.

matrei committed:
build: remove unused `JavadocFixTool` from `buildSrc`

matrei committed:
build: move dependency declaration to projects

Each project should explicitly declare its own dependencies
for better clarity and maintainability.

matrei committed:
build: move dependency declaration to projects

Each project should explicitly declare its own dependencies
for better clarity and maintainability.

matrei committed:
build: move dependency declaration to projects

Each project should explicitly declare its own dependencies
for better clarity and maintainability.

matrei committed:
build: extract Java config and add `javaVersion` property

Move shared Java configuration to a separate Gradle file
for better maintainability and reuse across projects.

Add project-wide `javaVersion` Gradle property.

matrei committed:
build: move plugin application to individual projects

Each project should explicitly declare its main Gradle plugins
for better clarity and maintainability.

matrei committed:
build: add Java configuration async projects

Add missing Java configuration to the async projects.

matrei committed:
build: clean up documentation generation

- Simplify `groovydoc` generation
- Disable `javadoc` generation
- Remove `asciidoc` generation

The purpose of downloading the external groovy-doc project
and generating documentation pages in this project was unclear,
so it has been removed for now.

matrei committed:
build: set `version` and `group` in individual projects

Move `version` and `group` declarations to each project
instead of relying on conditional logic in the root build file.
This improves clarity and ensures explicit configuration per project.

matrei committed:
build: clean up root `assemble` task

matrei committed:
build: streamline functional tests Gradle plugin usage

- Remove unnecessary explicit application of `groovy`, `java`,
  `war`, and `application` plugins.
- Exclude `gsp` and `gson-views` pre-compilation in functional
  test apps (only needed for functional test plugins).
- Retain the `asset-pipeline` Gradle plugin for runtime asset
  compilation but disable the `assetCompile` task to avoid
  unnecessary pre-compilation.

matrei committed:
build: remove unused file `integration-test.gradle`

matrei committed:
build: extract functional test configuration

matrei committed:
build: extract assemble configuration

matrei committed:
build: add `groovy` and `jar` task config to `java-config`

matrei committed:
build: extract test config

matrei committed:
build: extract publish config

matrei committed:
build: adapt shell cli deps configuration name

The name was changed in `assemble-config` in a previous commit

matrei committed:
build: remove unused `findbugs.gradle`

matrei committed:
build: add fix for `grails-dependencies` publishing

Grails Gradle Publish Plugin requires a non-empty sourceSet present

matrei committed:
build: remove `javadoc` config

This configuration is handled by the
Grails Gradle Publish plugin now.

matrei committed:
build: add skip check to `test-suite-uber`

jdaugherty commented on gradle/assemble-config.gradle:
```diff
@@ -0,0 +1,19 @@
+tasks.withType(PublishToMavenLocal).configureEach {
+    dependsOn('installToHomeDist')
+    doLast {
+        copy {
```
From what I can tell, `homeDistDir` is set to `dist` in the root project directory.

So the same concern I have with installToHomeDist, I have here.  What ensures the changed dependencies are actually repopulated and old jars aren't left in that directory?  I don't see a clean-up of this directory.

matrei replied:
These distribution directories are deleted by a `clean` task in the root build file.

jdaugherty commented on gradle/docs-config.gradle:
```diff
@@ -0,0 +1,19 @@
+tasks.named('groovydoc', Groovydoc) {
+    classpath = configurations.documentation
+    groovyClasspath = configurations.documentation
+    access = GroovydocAccess.PRIVATE
```
We have been changing these to protected in other projects, can we change it here?

jdaugherty commented on gradle/docs-config.gradle:
```diff
@@ -0,0 +1,19 @@
+    includeInApiDocs = true
+}
+
+tasks.named('groovydoc', Groovydoc) {
```
grailsPublish ties groovydoc and javadoc as long as you have this: 

	java {
	    withSourcesJar()
	    withJavadocJar()
	}

Why is this needed here?

matrei replied:
When removing this and adding `java.withJavadocJar()` the Grails Gradle Plugin seems to package both javadoc and groovydoc so that when unzipping the javadoc jar it has duplicates of the files, so you cannot extract it. And in the `build/docs` directory there is a `groovydoc` and a `javadoc` dir.

Shouldn't we only run `groovydoc` and package that into the `javadoc` jar, like I have done in this project now?

jdaugherty replied:
This is intentional.  if you unzip with a commandline client it will overwrite the files and it will be as you expect.  We opened a bug ticket on gradle to make it replace instead of add duplicates.

I believe the reason we don’t run just the groovydoc package is some projects do include java only in a java source.  

As for the directories, our intention was to build the jar.  I need to look at the directories to see whats happening.

matrei replied:
In the `grails-core` repo, there is not one single `src/main/java` source set.
Which projects are we using `src/main/java` in?

I don't think the solution is optimal at the moment. We are effectively generating and packaging both groovydoc and javadoc only to throw away the groovydoc when extracting the jar.

Should we be advocating groovy usage, and package generated groovydoc in the doc jar and only use the `src/main/groovy` source set?

matrei replied:
> Can you please confirm you've published to maven local before this change & saved off the artifacts. Then, published after this change and diffed the directory structure / names / files? This will ensure we haven't broken anything.

Confirmed (after some fixes, thank you!)

jdaugherty replied:
I'm only suggesting we fix this in the gradle plugin in a separate pull and switching to this for now to be consistent across the projects.

matrei replied:
OK, I'll switch to letting the Gradle Plugin handle this.

jdaugherty replied:
https://github.com/apache/grails-gradle-plugin/issues/430

jdaugherty commented on gradle/java-config.gradle:
```diff
@@ -0,0 +1,19 @@
+    options.forkOptions.jvmArgs = ['-Xms128M', '-Xmx1G']
+}
+
+tasks.withType(Jar).configureEach {
```
Why keep the attribute addition when we build in github actions and no other 'core' project does this?

matrei replied:
I'm don't know why it is added, so I did not want to remove it. Maybe we can remove it?

jdaugherty replied:
lets remove it; its likely left over pre-github actions change

matrei replied:
I do not understand the connection between adding these attributes to the jar files and using GitHub actions.
Are you saying that by using GitHub actions, this is not necessary?

jdaugherty replied:
Yes, I am saying that historically grails was built on travis and before that on someone’s machine so it was important to know who built it.  Now it’s not needed because we have a standard CI.

matrei replied:
OK, thanks for clearing that up for me, I'll remove it.

matrei replied:
It seems removing the jar attributes broke `GrailsUtilTest` which is testing `GrailsUtil.getGrailsVersion()` which seems to be fetching the version from the jar attributes.

jdaugherty replied:
is this there for the shell then?  or is that test even relevant now?

matrei replied:
> is this there for the shell then? or is that test even relevant now?

`GrailsUtil.getGrailsVersion()` is used in 19 places in the `grails-core` projects.

jdaugherty replied:
So it looks like Environment.groovy is what uses this.  That's fine, but can we apply this only to the bootstrap class then?  It looks for the manifest file in the same location as the Environment.class, which is only in the bootstrap project.

matrei replied:
> but can we apply this only to the bootstrap class then?

Do you mean add the version to the jar manifest only in the `grails-bootstrap` project?

matrei replied:
After moving the attributes to only the `grails-bootstrap` jar I'm getting test errors all over the place.
Seems to be originating from:
https://github.com/apache/grails-data-mapping/blob/00fbc39170146c6ae844c66c7e2056a4e7565ba6/grails-datastore-core/src/main/groovy/org/grails/datastore/mapping/core/grailsversion/GrailsVersion.groovy#L147

matrei replied:
We should keep the JAR attributes in this PR to ensure functionality is preserved. We can revisit once Snapshots are up and running.

jdaugherty replied:
I agree we should address after this PR.  We should consolidate to one, but it's out of scope of this PR.

https://github.com/apache/grails-core/issues/14098

jdaugherty commented on gradle/unit-test.gradle:
```diff
@@ -1,129 +0,0 @@
-// todo Unify test tasks into one multithreaded execution unit with a custom fork frequency
-// todo Add test progress listener with dot notation.
-
-configurations {
```
We probably should create a ticket to adopt jacoco and setup a github actions to publish code coverage %

matrei replied:
#14096

jdaugherty commented on grails-dependencies/src/main/resources/publish-fix:
```diff
@@ -0,0 +1 @@
+Grails Gradle Plugin needs something in a sourceSet to be able to publish
```
This is fine for this review, but I would like to fix this separately:

What do you think about fixing this by adding a 'bom' property to grails publish?  It would remove this requirement and be a way to convey this project only publishes dependencies.  I am open to other names for the property if you can think of one.

matrei replied:
Yes, that would make sense.

jdaugherty replied:
https://github.com/apache/grails-gradle-plugin/issues/429

jdaugherty commented on grails-docs/build.gradle:
```diff
@@ -1,10 +1,23 @@
+    implementation platform(project(':grails-bom'))
+
     gradleConf gradleApi()
+    api "org.apache.groovy:groovy:$groovyVersion"
```
Will this work?  This is a gradle task so it must use the groovy version of gradle.

matrei replied:
IDK. I just refactored this dependency to be added in the subproject build file instead of in the root build file.
The `grails-docs` was not in the `compiledByGradleGroovyVersion` collection.
https://github.com/apache/grails-core/blob/f673a05c5e1e7e61830608564c0497554eaefbf4/build.gradle#L63
https://github.com/apache/grails-core/blob/f673a05c5e1e7e61830608564c0497554eaefbf4/build.gradle#L401-L410
https://github.com/apache/grails-core/blob/f673a05c5e1e7e61830608564c0497554eaefbf4/build.gradle#L598-L604

jdaugherty replied:
I think we have found a bug =)  

Can we please change this to be the groovy version that's used by gradle?

jamesfredley replied:
I think we got lucky on this one and the Groovy 4 compiled classes ran fine in Groovy 3, but I do agree with changing any code executed during the Gradle build to use `$GroovySystem.version` as a precaution.

matrei replied:
Changed!

jdaugherty commented on grails-logging/build.gradle:
```diff
@@ -1,3 +1,41 @@
+
+    implementation platform(project(':grails-bom'))
+
     api project(":grails-core")
```
single quotes?

matrei replied:
:smile: I had that one coming! However, this line has not been touched in this PR.

I have NOT gone through the dependencies sections in this PR, as I know the PR is blocking progress and I want to get it done.

Clarifying and cleaning up dependencies can be done in incremental steps later.
There is also work left to clean up the rest of the subproject build files.

jdaugherty replied:
Makes sense.  Thank you.

jdaugherty commented on grails-plugin-codecs/build.gradle:
```diff
@@ -1,4 +1,50 @@
+
+    implementation platform(project(':grails-bom'))
+
     api project(":grails-web"), project(":grails-encoder")
```
I have never seen this syntax; can this be split across 2 lines?

Single quotes?

matrei replied:
See previous response

jdaugherty commented on grails-plugin-databinding/build.gradle:
```diff
@@ -1,5 +1,18 @@
+
+    implementation platform(project(':grails-bom'))
+
     api project(':grails-core'), project(':grails-web')
```
Can we please split these across multiple lines?

matrei replied:
See previous response

jdaugherty commented on grails-plugin-i18n/build.gradle:
```diff
@@ -1,4 +1,50 @@
+
+    implementation platform(project(':grails-bom'))
+
     api project(":grails-web")
```
single quotes?

matrei replied:
See previous response

jdaugherty commented on grails-plugin-url-mappings/build.gradle:
```diff
@@ -1,4 +1,50 @@
+
+    implementation platform(project(':grails-bom'))
+
     api project(":grails-web"), project(":grails-plugin-controllers")
```
split & single quotes?

matrei replied:
See previous response

jdaugherty commented on grails-plugin-validation/build.gradle:
```diff
@@ -1,4 +1,50 @@
+    implementation platform(project(':grails-bom'))
+
     api project(":grails-core"),
         project(':grails-web')
```
can we declare the grails-web dependency as a separate api dependency?

matrei replied:
See previous response

jdaugherty commented on grails-spring/build.gradle:
```diff
@@ -1,7 +1,44 @@
+
+    implementation platform(project(':grails-bom'))
+
     api "org.springframework:spring-tx"
```
single quotes?

matrei replied:
See previous response

jdaugherty commented on grails-test-suite-base/build.gradle:
```diff
@@ -1,4 +1,16 @@
+
+    implementation platform(project(':grails-bom'))
+
     api project(':grails-bootstrap'),
```
can we split this up into separate api statements?

matrei replied:
See previous response

jdaugherty commented on grails-test-suite-base/build.gradle:
```diff
@@ -7,4 +19,35 @@ dependencies {
         project(':grails-plugin-domain-class')
 
     api project(":grails-plugin-converters")
```
single quotes?

matrei replied:
See previous response

jdaugherty commented on grails-test-suite-persistence/build.gradle:
```diff
@@ -1,3 +1,12 @@
+group = 'org.grails'
+
 configurations.testCompileClasspath {
     exclude module: "grails-plugin-testing"
```
single quotes?

matrei replied:
See previous response

jdaugherty commented on grails-test-suite-persistence/build.gradle:
```diff
@@ -1,3 +1,12 @@
+
 configurations.testCompileClasspath {
     exclude module: "grails-plugin-testing"
 }
```
Can we split the dependencies into their own lines and scopes?

matrei replied:
See previous response

jdaugherty commented on grails-test-suite-uber/build.gradle:
```diff
@@ -1,39 +1,60 @@
+
+
+    // Testing
+    testImplementation('org.spockframework:spock-core') { transitive = false }
```
Do you know why we include the transitive other places but exclude it here?

matrei replied:
No, see previous response

jdaugherty commented on grails-web-mvc/build.gradle:
```diff
@@ -1,4 +1,50 @@
+
+    implementation platform(project(':grails-bom'))
+
     api project(":grails-web-common"),
```
Can we split these?

matrei replied:
See previous response

jdaugherty commented on grails-web-url-mappings/build.gradle:
```diff
@@ -1,7 +1,20 @@
+
+    implementation platform(project(':grails-bom'))
+
     api project(":grails-web-common")
```
single quotes?

matrei replied:
See previous response

jdaugherty commented on grails-web/build.gradle:
```diff
@@ -1,8 +1,54 @@
     api project(":grails-web-mvc")
+    api "org.apache.groovy:groovy:$groovyVersion"
 
     api "org.grails:grails-web-gsp"
```
single quotes?

matrei replied:
See previous response

jdaugherty commented on grails-web/build.gradle:
```diff
@@ -1,8 +1,54 @@
     api project(":grails-web-mvc")
+    api "org.apache.groovy:groovy:$groovyVersion"
 
     api "org.grails:grails-web-gsp"
```
side topic: isn't this in the grails views project?  This seems like a reason to bring views into core

matrei replied:
`grails-web-gsp` is in fact not used in `grails-web` and can be removed (unless some downstream project depends on `grails-web` providing it, which we know is a bad idea).

It seems `grails-plugin-mimetypes` is using `org.grails.gsp.ResourceAwareTemplateEngine` and it's sole dependency is `grails-web`.

jdaugherty replied:
Looks like ResourceAwareTemplateEngine is used in grails-web-common too.  I suspect we need to revisit the project structure of gsp.  It seems like over time several classes have been pulled up the project chain.  I'll take a look at this after the apache publishing is working.

jdaugherty replied:
https://github.com/apache/grails-views/issues/726

jdaugherty commented on build.gradle:
```diff
@@ -36,569 +10,66 @@ ext {
             metaInf {
-                ignoreAttribute("Built-By")
-                ignoreAttribute("Created-By")
+                ignoreAttribute('Built-By')
```
Let's assume we remove the properties,  then would this still be required?  Can we remove this + remove the jar properties?

jdaugherty replied:
I think we can remove this if we only apply it to bootstrap.  It's ok to miss the cache on that library.

matrei replied:
We should keep this and the JAR attributes in this PR to ensure functionality is preserved. We can revisit once Snapshots are up and running.

jdaugherty replied:
https://github.com/apache/grails-core/issues/14098

jdaugherty commented on gradle/assemble-config.gradle:
```diff
@@ -0,0 +1,19 @@
+tasks.register('installToHomeDist', Copy) {
```
From what I can tell, `distInstallDir` is set to `build/dist-tmp` in the root project.  

This is included in each project, so that means each project will copy it's jars into this directory.  That means we can't use the `Sync` task to "cleanup" what's not supposed to belong.  That's unfortunate: if the dependencies change, how do we ensure the old dependency isn't still copied to build/dist-tmp?  I don't see any code that forces a clean when these tasks are run across projects.

matrei replied:
Won't the `inputs` of the `installToHomeDist` task make sure any changes to the output jars are detected and that they are copied to/overwritten in `distInstallDir`?

jdaugherty commented on gradle/assemble-config.gradle:
```diff
@@ -0,0 +1,19 @@
+tasks.register('installToHomeDist', Copy) {
+    inputs.files(tasks.named('jar').map { it.outputs.files })
```
Isn't what determines the input to this task what creates the libs?  Thus, it's just the configuration that's needed? Shouldn't we let gradle determine the files by the configuration instead of explicitly listing them all?

matrei replied:
I don't think using only the dependency configuration will do it. There are other factors (like changed sources) that can change the `jar` task output.

jdaugherty commented on gradle/assemble-root-config.gradle:
```diff
@@ -5,40 +5,23 @@ import org.grails.gradle.GrailsBuildPlugin
-
-            for (artifact in configuration.resolvedConfiguration.resolvedArtifacts) {
+        project(':grails-shell').configurations.cli.with {
+            exclude(group: 'org.codehaus.groovy')
```
Shouldn't this exclude `org.apache.groovy` too?

jdaugherty replied:
Since these are compiled with the groovy version of gradle, this is ok for now.

jdaugherty commented on gradle/assemble-root-config.gradle:
```diff
@@ -57,16 +40,20 @@ tasks.register('populateDependencies', Sync) {
-        !subproject.name.startsWith('grails-bom') &&
-        !subproject.name.contains('grails-test-examples')
+    from subprojects.findAll {
+        !it.name.startsWith('grails-test-suite') &&
```
What are your thoughts about adding a property to the root that captures the list of projects that are published but not a test / bom / dependency set?  It seems like this logic is still used after the refactor so having it in one place would make this easier.

It seems this set of projects is used more than once in this file alone.

matrei replied:
Done!

jdaugherty commented on gradle/assemble-root-config.gradle:
```diff
@@ -116,7 +103,7 @@ class GrailsCreateStartScripts extends CreateStartScripts {
+    dependsOn 'syncCliDependencies', 'grailsCreateStartScripts'
     subprojects { Project project ->
         if(!project.name.startsWith('grails-test-suite') && !project.name.startsWith('grails-test-examples')) {
             task.dependsOn("$project.name:publishToMavenLocal")
```
Shouldn't this be the same set of projects from the line#44 comment?

matrei replied:
Refactored

jdaugherty commented on gradle/assemble-root-config.gradle:
```diff
@@ -126,7 +113,8 @@ tasks.register('install') { task ->
 //task install(dependsOn: [populateDependencies, grailsCreateStartScripts] + subprojects.findAll { !it.name.startsWith('grails-test-suite') }
 //                                                                                                        *.collect { Project p -> p.tasks.withType(PublishToMavenLocal)})
```
Can we removed the commented code?

matrei replied:
Yes

jdaugherty commented on gradle/functional-test-config.gradle:
```diff
@@ -0,0 +1,39 @@
+        .findAll { it.group == 'org.grails.plugins' }
+        .collect { it.name }
+
+configurations.configureEach {
```
Can we leave the original comment here (slightly adjusted)?  

        // Test projects will often include dependencies from grails-core, this will ensure any dependencies included
        // will be substituted with projects in this repository instead of pulling upstream

matrei replied:
Oh, sorry, I did not mean to remove that. I'll add it back.

jdaugherty commented on gradle/functional-test-config.gradle:
```diff
@@ -0,0 +1,39 @@
+            for (def possibleProject : rootProject.subprojects) {
+                if (!possibleProject.name.startsWith('grails-test-suite') && !possibleProject.name.contains('grails-test-examples')) {
+                    def artifactId = mappedArtifactIds[possibleProject.name] ?: possibleProject.name
+                    def substitutedArtifact = "${possibleProject.name in pluginsGroupProjects ? 'org.grails.plugins' : 'org.grails'}:$artifactId"
```
Now that we have the projects specifying `pomArtifactId` and the group is defined inside of the project, do we even need this mapping?  Can't this now be `${possibleProject.group}:${possibleProject.pomArtifactId ?: possibleProject.name}`?

matrei replied:
For some reason `grails-testing-support` and `grails-web-boot` do not show the right `group`.
Added some logging:
```console
Substituting org.grails:grails-spring with project grails-spring
Substituting org.grails:grails-test with project grails-test
Substituting grails.core.ROOT:grails-testing-support with project grails-testing-support
Substituting org.grails:grails-web with project grails-web
Substituting grails.core.ROOT:grails-web-boot with project grails-web-boot
Substituting org.grails:grails-web-common with project grails-web-common
```

So, as it is now, it only works as they are in the default group.

jdaugherty replied:
If you add the substitution code in a ` project.afterEvaluate { } ` block, does it work?

matrei replied:
> If you add the substitution code in a project.afterEvaluate { } block, does it work?

No, that does not seem to work.

jdaugherty replied:
Making the functional tests dependent on the framework projects seems to fix resolving the group:  

		Set<Project> nonFunctionalProjects = rootProject.subprojects.findAll { !it.name.startsWith('grails-test-examples') }
		nonFunctionalProjects.each {
		    project.evaluationDependsOn(it.path)
		}

jdaugherty commented on gradle/functional-test-config.gradle:
```diff
@@ -0,0 +1,39 @@
+        dependencySubstitution {
+            for (def possibleProject : rootProject.subprojects) {
+                if (!possibleProject.name.startsWith('grails-test-suite') && !possibleProject.name.contains('grails-test-examples')) {
+                    def artifactId = mappedArtifactIds[possibleProject.name] ?: possibleProject.name
```
Is this mapping still needed?  Each project will specify it directly now, yes?

jdaugherty commented on grails-bom/build.gradle:
```diff
@@ -15,6 +19,9 @@ def versions = new Properties()
 versions.load(new StringReader(new File("$projectDir.parentFile/gradle.properties").text))
 
 ext {
```
Can we update the task registrations in this file to be lazy?

jdaugherty reviewed (changes requested):
1. as discussed, we should leave the gradle plugins (gsp, views, etc) in the functional tests. I realize that the compiled gsps aren't used by the tests because we run in dev mode, but it still exercises that plugin in a basic manner.  Eventually, we'll want to add production mode tests to make use of these for end-to-end coverage.

2. Can you please confirm you've published to maven local before this change & saved off the artifacts.  Then, published after this change and diffed the directory structure / names / files?  This will ensure we haven't broken anything.

3. Can you please confirm you've made sure the same # of tests run before and after this change? 

See other comments, but overall this is an amazing job.  Thank you so much!

matrei committed:
build: fix root project delete task

jdaugherty commented:
Did you also test grails-doc and make sure if it has this code that it can still generate?

matrei committed:
build(docs): pr feedback

Change to `groovydoc.access` to `PROTECTED`.

matrei committed:
chore: pr feedback - remove commented code

matrei committed:
build: pr feedback - add back relevant comment

matrei committed:
build: pr feedback - simplify test dependency substitution

matrei committed:
build: pr feedback - add back gradle plugins to functional test

matrei committed:
build: pr feedback - remove jar adding jar attributes

matrei committed:
Revert "build: pr feedback - remove jar adding jar attributes"

This reverts commit 30044a5fc393610c5d02b348523ace6bdbb41bbb.

matrei committed:
build: pr feedback - lazy task registration

matrei committed:
build: add `pomArtifactId` to `events` and `converters` plugins

matrei committed:
docs: fix javadoc linting errors

matrei committed:
ci: Update Develocity plugins and settings

matrei committed:
ci: Fix Develocity settings

matrei committed:
build: Fix test reporting in `test-suite-web`

matrei committed:
build: Fix tests integration tests not running

matrei commented:
> Can you please confirm you've made sure the same # of tests run before and after this change?

Confirmed `2298 tests (28 ignored)` (after a fix, thank you!)

matrei commented:
> Can you please confirm you've published to maven local before this change & saved off the artifacts. Then, published after this change and diffed the directory structure / names / files? This will ensure we haven't broken anything.

Confirmed (after some fixes, thank you!)

matrei commented:
> Did you also test grails-doc and make sure if it has this code that it can still generate?

I don't know what you mean by this.

jdaugherty commented:
> > Did you also test grails-doc and make sure if it has this code that it can still generate?
> 
> I don't know what you mean by this.

https://github.com/apache/grails-doc can either checkout or point to a grails home.  It will then build the documentation to be published.  You should be able to checkout that project, point it to this branch, and then run the doc task to generate documentation.  If it doesn't work, we'll need to adjust the gradle files ...

matrei committed:
build: Simplify functional test dependency substitution

matrei committed:
build: Use Gradle plugin to package javadoc jar

matrei committed:
build(windows): Fix javadoc error

error: unmappable character (0x90) for encoding

matrei commented:
> > > Did you also test grails-doc and make sure if it has this code that it can still generate?
> > 
> > 
> > I don't know what you mean by this.
> 
> https://github.com/apache/grails-doc can either checkout or point to a grails home. It will then build the documentation to be published. You should be able to checkout that project, point it to this branch, and then run the doc task to generate documentation. If it doesn't work, we'll need to adjust the gradle files ...

Aha, I see what you mean. It does seem to work. `grails-doc` even includes Groovy 4 on the build classpath!!

https://github.com/apache/grails-doc/blob/174051c0d071bc889d213f46820d2aab9b02eb4a/buildSrc/build.gradle#L16-L20

matrei committed:
build: compile `grails-docs` with Groovy 3

grails-docs classes are used in Gradle builds, so we
must compile with Groovy 3 to ensure compatibility
(until Gradle upgrades to Groovy 4).

matrei committed:
build: extract install project filter to method

matrei committed:
build: remove `sourcesJars` task

The source jars are already included in the `dist` directory
in the distribution. No need to duplicate them also in the `src`
directory.

Previously, for some reason, also `grails-datastore*` source jars
where included in the distribution `src` directory, but after
recent changes that was no longer the case.

matrei committed:
build: remove `homeSrcDir` as it is no longer used

matrei committed:
build: dry up test projects condition check

jdaugherty commented on build.gradle:
```diff
@@ -35,570 +9,76 @@ ext {
-                stagingProfileId = ossStagingProfileId
-            }
-        }
+    mappedArtifactIds = [
```
Should we build this list based on the pomArtifactId? It looks like this is only used in the grails bom.  Maybe move the logic there?

matrei replied:
Thanks for reminding me!

jdaugherty commented on build.gradle:
```diff
@@ -35,570 +9,76 @@ ext {
-                        }
-                    }
+                if (details.requested.group == 'org.spockframework') {
+                    details.useVersion(findProperty('spock.version') as String)
```
Since we have a bom, is this required?

matrei replied:
Good catch!

jdaugherty commented on gradle/docs-config.gradle:
```diff
@@ -0,0 +1,14 @@
+    includeInApiDocs = true
+}
+
+tasks.named('groovydoc', Groovydoc) {
```
Shouldn't this be .configure { }?

matrei replied:
I don't think that should not be necessary when using `tasks.named()`.

jdaugherty commented on grails-core/src/main/groovy/grails/core/DefaultGrailsApplication.java:
```diff
@@ -627,8 +627,8 @@ public ArtefactInfo getArtefactInfo(String artefactType) {
      * @return The return value of the method
-     * TODO Need to add matches for add<Artefact>Class(java.lang.Class) and add<Artefact>Class(GrailsClass)
      */
+    // TODO Need to add matches for add<Artefact>Class(java.lang.Class) and add<Artefact>Class(GrailsClass)
```
By moving this comment out of the javadoc, won't this mean it won't show externally in the api documentation?

matrei replied:
Yes, that is true. I did not recognize it as part of the javadoc, but more of a code comment. Do you think it should be part of public javadoc?

jdaugherty replied:
I'm not sure, maybe since it was there before it was intentional?  I'd leave it for now.

jdaugherty commented on grails-docs/build.gradle:
```diff
@@ -1,24 +1,51 @@
+
+    // grails-docs classes are used in Gradle builds,
+    // so we must compile with Groovy 3 until Gradle upgrades to Groovy 4.
+    compileOnly "org.codehaus.groovy:groovy:$GroovySystem.version"
```
thank you!

jdaugherty commented on grails-test-examples/hyphenated/build.gradle:
```diff
@@ -11,10 +11,6 @@ version "0.1"
 group "hyphenated"
 
 apply plugin: "org.grails.grails-web"
-apply plugin: "org.grails.grails-gsp"
```
Looks like this project has gsps, can we keep the gsp plugin?

matrei replied:
The demo33 project has gsp compilation check during build (with explaining comments).
I thought it would be enough to have that smoke test in one project.

jdaugherty commented on grails-test-examples/issue-11102/build.gradle:
```diff
@@ -11,18 +11,8 @@ buildscript {
-apply plugin: "war"
 apply plugin: "asset-pipeline"
 apply plugin: "org.grails.grails-web"
-apply plugin: "org.grails.grails-gsp"
```
Looks like the project still has gsps, can we keep this?

matrei replied:
See previous comment.

jdaugherty commented on grails-test-examples/issue-views-182/build.gradle:
```diff
@@ -4,18 +4,13 @@ buildscript {
 group "issueviews182"
 
 apply plugin:"org.grails.grails-web"
-apply plugin:"org.grails.plugins.views-json"
```
There are `gson` views in this project, can we please keep this?

matrei replied:
The demo33 project has json-views compilation check during build (with explaining comments).
I thought it would be enough to have that smoke test in one project.

jdaugherty commented on grails-test-examples/micronaut/build.gradle:
```diff
@@ -4,24 +4,16 @@ buildscript {
-apply plugin:"idea"
-apply plugin:"war"
 apply plugin:"org.grails.grails-web"
-apply plugin:"org.grails.grails-gsp"
```
Can we keep gsp?

matrei replied:
See previous comment.

jdaugherty reviewed (approved):
Can we please be sure to add back the remaining gsp / gson plugins where I marked?

matrei committed:
build: refactor `mappedArtifactIds` to use `pomArtifactId`

matrei committed:
build: remove explicit `resolutionStrategy` for `spock`

The version should be resolved from `grails-bom` now.

matrei commented:
> Can we please be sure to add back the remaining gsp / gson plugins where I marked?

We could, but I think we are only wasting build time. The demo33 project smoke tests the Gradle plugins.

jdaugherty commented:
> > Can we please be sure to add back the remaining gsp / gson plugins where I marked?
> 
> We could, but I think we are only wasting build time. The demo33 project smoke tests the Gradle plugins.

We don't know how these projects would diverge over time so it's safe to add them to all of them.  Specifically, micronaut is likely going to be different than the base grails app so it's probably critical to have it there at the least.

jamesfredley commented on grails-core/build.gradle:
```diff
@@ -1,10 +1,23 @@
     api "jakarta.annotation:jakarta.annotation-api"
 
     implementation "com.github.ben-manes.caffeine:caffeine"
+    api "org.apache.groovy:groovy:$groovyVersion"
```
These surprised me, but they were in the POMs and did not cause version issues so we are fine.

jdaugherty replied:
the hard coded version is likely why we have to use the override technique we do for groovy.  Since it already did this, I wasn't going to advocate for change. But eventually, we probably should pull that version from the bom instead.

jamesfredley commented:
I tested grails-shell CLI zip and these changes add a doc directory with groovydocs (1,815 files) whereas it was 222 files total before.  I think we should skip the docs in the zip distribution.  

Additionally this issue https://github.com/apache/grails-core/pull/14065/files is back on Windows.

jdaugherty commented:
> I tested grails-shell CLI zip and these changes add a doc directory with groovydocs (1,815 files) whereas it was 222 files total before. I think we should skip the docs in the zip distribution.
> 
> Additionally this issue https://github.com/apache/grails-core/pull/14065/files is back on Windows.

I guess we should skip the withJavadoc() for the cli project?

jamesfredley reviewed (approved):
I'd like to fix the grails-shell CLI issue on Windows that reappeared before merging this, but would also be OK moving that to another ticket.

matrei committed:
build: fix regression of windows cli path length problem

It seems dependency configuration excludes do not affect
inherited dependencies.

Related: gh-14065

matrei commented:
> I tested grails-shell CLI zip and these changes add a doc directory with groovydocs (1,815 files) whereas it was 222 files total before. I think we should skip the docs in the zip distribution.

Are you sure this has changed? I checked the previous release ([v7.0.0-M3](https://github.com/apache/grails-core/releases/download/v7.0.0-M3/grails-7.0.0-M3.zip)), and it appears to include the API docs in the doc subdirectory.

matrei committed:
build: fix doc api subdirectory in distribution

This aligns with previous zip distribution releases.

matrei committed:
build: smoke test production view compilation in functional tests

This commit adds back the 'org.grails.grails-gsp' and
`org.grails.plugins.views-json` Gradle Plugins to the functional tests.

This will act as a smoke test that production compilation of views
does not fail when building.

jamesfredley reviewed (approved):
@matrei The changes for the CLI on windows work perfectly.

I went back and checked and you are correct on the doc directory being present for 7.0.0-M3 and 7.0.0-M1.  The doc directory was not present on 6.2.x distributions and I have created an issue to explore that separately.  https://github.com/apache/grails-core/issues/14100

Great work reorganizing this build!!!
